### DISCUSSION TITLE
Subitem in a form

### DISCUSSION THREAD

**Discussion by mattiason** [Budibase/budibase#2043] [Help] [answered]
If i a have a form eg a quote is there anyway that it is possible to add line items in the form view to table via button that automatically created a subitem in a related table that is show in the table?

**Comment by mjashanks** [comment, marked as answer]
Hey @mattiason.

Below is a video showing how I have achieved this, starting with a demo. Let me know if I have not understood you.

Key points

1. I have a Customers table and a linked Notes table. A customer has Many notes

On my Customer edit screen...

2. I add a new form, below the customer form, for adding a new note - here I set the schema of the form to `Notes`
3. In my "Add Note" button I set the Note.Customer property to the `Url.id` i.e. Id of the customer that we are currently editing. The data source is set to me new notes form, and the table is et to `Notes`
4. I then add a new data provider, set to `Relationships > Customer.Notes`
5. I add a table inside this Data Provider. This makes the table display notes that belong to that customer.


https://user-images.githubusercontent.com/3524181/126523241-862417ca-275c-4840-9d11-b163a2d8a4e4.mp4

**Reply by mjashanks** [reply]
It's most likely to do with the nesting of your data providers.

It should look like ...

`Data Provider (customer) > Repeater > Data Provider (Customers.Notes)`


There will be other components in there of course, but that should be the hierarchy of data providers.

**Reply by KernOil** [reply]
@mjashanks  my notes look like this why is that?
![image](https://user-images.githubusercontent.com/80058238/129420179-33ac144e-297c-4727-814e-3bcd05a8d5d6.png)

**Reply by mjashanks** [reply]
Hey @AndrewJRamay - what's in your `Save Row` action, inside the "Save Note" button?

**Reply by KernOil** [reply]
Hi @mjashanks 
this is what my button does, 
![image](https://user-images.githubusercontent.com/80058238/129583792-872659c3-9454-4b14-b54e-031db610c81a.png)

they show up just fine on this screen but not when you click into details for the row 
![image](https://user-images.githubusercontent.com/80058238/129584087-c60548c0-caa7-422e-9d60-fe24ed02a259.png)

**Reply by mjashanks** [reply]
Are you using a BudibaseDB table Or SQL?

**Reply by KernOil** [reply]
@mjashanks I am using all BB for this app.

**Reply by KernOil** [reply]
it will save the row properly and have no errors. 
![image](https://user-images.githubusercontent.com/80058238/129588613-bc43a787-e242-4f85-bf25-ae26e40a53c6.png)

**Reply by KernOil** [reply]
@mjashanks I can send you the app if you want to look at it.

**Reply by mjashanks** [reply]
Yea - that'd be good thanks - I think you have my email!

**Reply by KernOil** [reply]
Yes, I should. I'll send it over.

**Reply by mjashanks** [reply]
Hey @AndrewJRamay - this was a weird one, took me a while to work it out. There is a bug here.

The issue is caused by your primary display column on the `Notes` table being called `Notes`. It conflicts with the table name somehow. 

Rename your column to anything else - e.g. `Text` and it works immediately.

**Reply by KernOil** [reply]
@mjashanks that is really interesting, thank you for solving that for me!

**Comment by mattiason** [comment]
Thank is working good, but is there any way to clear the form automatically

**Reply by mjashanks** [reply]
Unfortunately not. It would be useful to have a `Set Value` or `Clear Form` action for this reason. Have logged here: https://github.com/Budibase/budibase/discussions/2059

**Comment by mattiason** [comment]
Hi Thanks

Would also be creat if there was an option for a line edit. Like add a row
and edit it afterwards.

Kind regards
Mattias

**Reply by mjashanks** [reply]
Yea, we had also talked once about allowing "Line Edit" via a modal. It's not quite "inline", but does the same job: https://github.com/Budibase/budibase/discussions/1199

**Comment by KernOil** [comment]
@mjashanks I can get this to work with a relational SQL DB however I have to have the user select the column it is relating to even tho I have the save button saving on {{URL.id}} could you take a second and see if I am doing something wrong or if this is a bug like in #2402.

this is my relationship. one request has many notes.
![image](https://user-images.githubusercontent.com/80058238/129936944-2f5bbdec-929c-401f-aaa3-60a68aec0545.png)

here is my layout.
![image](https://user-images.githubusercontent.com/80058238/129937088-8f5ca7b0-5815-4b7c-becf-8828677c84aa.png)

here is my saved button. I have it saving to the FK because if I save it to the column BB makes as the key it will not save.
![image](https://user-images.githubusercontent.com/80058238/129950003-c4f3a8b4-5623-459b-99cc-154ac3224210.png)

in theory, I should be able to do this below because if I manually select the CAD Request column in the UI it will save but it will no do it if I call the row. I think there is a bug here.
![image](https://user-images.githubusercontent.com/80058238/129950306-5f237b48-7a59-49cc-a780-5663bf7433f0.png)

**Comment by mjashanks** [comment]
Hey Andrew. In your `Save Row` action...
1. Don' try and update `CAD Requests 5` - this is just used for displaying the linked row. Just delete this entry
2. For `request_fk`, instead of `{{ URL.id }}` use `{{ Repeater.requests.requests_id }}` (or whatever the ID column of your requests table is called.

In debugging this, I realised that my `{{ URL.id }}` = `[10]`, when I need it to be `10`. 

![image](https://user-images.githubusercontent.com/3524181/130036932-7fdd3571-4591-4f0e-93dd-f0c7fa74c8ed.png)

`%5B10%5D` is a URL encoded version of `[10]`

I'm not sure what this means - if it is a bug or not, but it's certainly confusing!

**Reply by mjashanks** [reply]
Interestingly, the `View` link on my customer table reads `/customers/{{ customers Table.customers._id }}`, which translates to `/customers/%5B10%5D`

However, if i browse to `/customers/10` the page works just fine.

Still, use the id from the repeater, as above to be safe.

**Reply by KernOil** [reply]
@mjashanks Thank you for checking this out I really appreciate it!

**Reply by KernOil** [reply]
@mjashanks  it appears to work however I am having trouble writing to the FK field it will save then the FK field will revert to null. did you have this problem?
I can see BB save the entry in the data view then BB will erase it.  any thoughts on to why that may be?

**Reply by KernOil** [reply]
@mjashanks here is a video of what I'm talking about for some reason it will not save to the FK unless the CAD request 5 columns is selected 

https://user-images.githubusercontent.com/80058238/130104440-c5845924-0d37-41a7-ae2a-077b92c737d7.mp4

**Reply by mjashanks** [reply]
Thanks Andrew. Try using `{{ Repeater.requests.NAME_OF_THE_ID_COLUMN }}` rather than `{{ Repeater.requests._id }}`

**Reply by KernOil** [reply]
@mjashanks ok, I tried every possible selection from the repeater values.  I keep getting the same outcome. the row will save but will be related to nothing.  I think the problem is the FK will not hold the value for whatever reason. it will only work if I save the value in the BB created relationship column 

so I would think that trying to save the BB-created relationship column with the correct value in the button would be the workaround but that will not allow the row to save.

https://user-images.githubusercontent.com/80058238/130110847-1786bc38-ea3e-4a1e-a6f7-aba40826b7b9.mp4

**Reply by KernOil** [reply]
the two columns I bring in from my DB is UserNOTES and REQUEST_FK

**Reply by mjashanks** [reply]
Hi @AndrewJRamay . I was looking at your "New Note" video.

I think your relationship is blank because you have a Relationship Picker inside your New Note Form, which you leave blank. I think the hardcoded value inside the save action should override, but it doesn't.

I was able to create this. As soon as I removed the relationship picker, it saved fine again.

**Reply by KernOil** [reply]
@mjashanks great observation! I have it working now thank you very much!

**Comment by astepke** [comment]
This is my first time writing in a forum. First of all congratulations on developing BB. I have tried several frameworks and with this I have been able to advance much more.
If it is possible that they make the same video as "Customer.With.Notes.App.mp4",
but starting from scratch. Thank you, have a nice day.

**Comment by arladmin** [comment]
Hi @mjashanks . I was wondering if you could guide me as well with my use-case with regards to the line-item adding capability.

I am setting up a Custom Form with several fields mapping back to various tables. I plan to save the entire data in their respective tables once the entire form has been filled. 

Is it possible to be able to show a variable number of these input fields based on the user's choice? 
(e.g. Imagine an Order Form. The user wants to attach a variable no. of Line items, from Table A, and another variable no. of Line items, from Table B, to that order form, and save it all in one final step only. In this case, intermediary saving is not possible as the Line Items need to be linked with the Order being worked on, but the Order will be generated only after the entire form has been filled.)
